### PR TITLE
feat: api proxy support - SQSERVICES-1794

### PR DESCRIPTION
**Pull request by netbe**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1794" title="SQSERVICES-1794" target="_blank">SubtaskSQSERVICES-1794</a>  Rework App Network Calls Order to account for Proxy
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Add SOCKS proxy support to network sessions.

### Causes (Optional)

Need to wait for proxy settings to be given by user (if any) before firing any requests.

### Solutions

Add the ability to pause network requests with ready property on transport sessions.

### Dependencies (Optional)

Needs releases with:

- [ ] https://github.com/wireapp/Starscream/pull/5

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Source/Public/ProxyCredentials.swift [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    public convenience init?(username: String?, password: String?, proxy: ProxySettingsProvider) {
+        guard let username = username, let password = password else {
+            return nil
+        }
+        self.init(proxy: proxy, username: username, password: password)
+    }
+
+    public func persist() throws {
+        guard let usernameData = username.data(using: .utf8),
+              let passwordData = password.data(using: .utf8) else { return }
```
Comment (David-Henner) [inline Source/Public/ProxyCredentials.swift:43]:
```suggestion
        guard 
           let usernameData = username.data(using: .utf8),
           let passwordData = password.data(using: .utf8) 
        else { return }
```

File: Source/PushChannel/StarscreamPushChannel.swift [hunk 2]
```diff
@@ -114,8 +119,19 @@ class StarscreamPushChannel: NSObject, PushChannelType {
...
-        webSocket = WebSocket(request: connectionRequest, certPinner: certificatePinning, useCustomEngine: true)
+        webSocket = WebSocket(request: connectionRequest, certPinner: certificatePinning, useCustomEngine: false)
         webSocket?.delegate = self
+
+        if let proxySettings = environment.proxy {
+            let proxyDictionary = proxySettings.socks5Settings(proxyUsername: proxyUsername, proxyPassword: proxyPassword).asDictionary()
+
+            let configuration = URLSessionConfiguration.default
+            configuration.connectionProxyDictionary = proxyDictionary
+            configuration.httpShouldUsePipelining = true
```
Comment (David-Henner) [inline Source/PushChannel/StarscreamPushChannel.swift:130]: Maybe you could move this in the implementation of `ProxySettingsProvider` to enable calling it this way 

```swift
webSocket?.configuration = proxySettings.urlSessionConfiguration
```

File: Source/TransportSession/UnauthenticatedTransportSession.swift [hunk 3]
```diff
@@ -77,21 +80,36 @@ final public class UnauthenticatedTransportSession: NSObject, UnauthenticatedTra
     private let userAgent: ZMUserAgent
     public var environment: BackendEnvironmentProvider
     fileprivate let reachability: ReachabilityProvider
-    
+    /// Property to accept requests
+    public let ready: Bool
```
Comment (David-Henner) [inline Source/TransportSession/UnauthenticatedTransportSession.swift:84]: Suggestion to rename it for better clarity

```suggestion
    public let readyForRequests: Bool
```

File: Source/TransportSession/UnauthenticatedTransportSession.swift [hunk 4]
```diff
@@ -77,21 +80,36 @@ final public class UnauthenticatedTransportSession: NSObject, UnauthenticatedTra
...
         
         super.init()
         
         let configuration = URLSessionConfiguration.ephemeral
         configuration.httpAdditionalHeaders = ["User-Agent": ZMUserAgent.userAgent(withAppVersion: applicationVersion)]
-        
+
+
+        if let proxySettings = environment.proxy {
+            let proxyDictionary = proxySettings.socks5Settings(proxyUsername: proxyUsername, proxyPassword: proxyPassword).asDictionary()
```
Comment (David-Henner) [inline Source/TransportSession/UnauthenticatedTransportSession.swift:107]: why don't we return the settings `asDictonary()` instead of calling it each time?

File: Source/TransportSession/UnauthenticatedTransportSession.swift [hunk 5]
```diff
@@ -128,17 +146,21 @@ final public class UnauthenticatedTransportSession: NSObject, UnauthenticatedTra
...
         return .success
     }
-    
-    @discardableResult
-    private func enqueueRequest(_ request: ZMTransportRequest) -> DataTaskProtocol {
+
+    private func enqueueRequest(_ request: ZMTransportRequest) {
+        guard ready else {
+            zmLog.info("Dropping request \(request) as networkTransportSession not ready")
+            // TODO: queue ZMTransportRequests until ready
```
Comment (David-Henner) [inline Source/TransportSession/UnauthenticatedTransportSession.swift:153]: is this going to be in another PR?

File: Source/TransportSession/ZMTransportSession.m [hunk 6]
```diff
@@ -163,61 +167,85 @@ - (instancetype)initWithEnvironment:(id<BackendEnvironmentProvider>)environment
...
-    ZMURLSession *foregroundSession = [[ZMURLSession alloc] initWithConfiguration:[[self class] foregroundSessionConfiguration] trustProvider:environment delegate:self delegateQueue:queue identifier:foregroundIdentifier userAgent:userAgent];
+
+    NSURLSessionConfiguration *foregroundConfiguration;
+    if (environment.proxy != nil) {
+        foregroundConfiguration = [[self class] foregroundSessionConfiguration];
+        foregroundConfiguration.connectionProxyDictionary = [proxyDictionary asDictionary];
+    }
+    else {
+        foregroundConfiguration = [[self class] foregroundSessionConfiguration];
+    }
```
Comment (David-Henner) [inline Source/TransportSession/ZMTransportSession.m:185]: If possible, I would move this to a helper method to keep the init method as readable as possible

File: Source/TransportSession/ZMTransportSession.m [hunk 7]
```diff
@@ -163,61 +167,85 @@ - (instancetype)initWithEnvironment:(id<BackendEnvironmentProvider>)environment
...
+        backgroundConfiguration = backgroundSessionConfiguration;
 
-    ZMTransportRequestScheduler *scheduler = [[ZMTransportRequestScheduler alloc] initWithSession:self operationQueue:queue group:group reachability:reachability];
+        backgroundConfiguration.connectionProxyDictionary = proxyDictionary;
+    }
+    else {
+        backgroundConfiguration = [[self class]
+                                   backgroundSessionConfigurationWithSharedContainerIdentifier:applicationGroupIdentifier
+                                   userIdentifier:userIdentifier];
+    }
```
Comment (David-Henner) [inline Source/TransportSession/ZMTransportSession.m:210]: same here

File: Tests/Source/ProxyCredentialsTests.swift [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        // GIVEN
+        let password = "12345"
+        let username = "testUsername"
+        let proxy = MockProxy(host: "testHost", port: 20, needsAuthentication: true)
+
+        // WHEN
+        let sut = ProxyCredentials(username: username, password: password, proxy: proxy)
+
+        // THEN
+        try sut?.persist()
```
Comment (David-Henner) [inline Tests/Source/ProxyCredentialsTests.swift:34]: what are we testing here? we're not asserting anything so I assume you just want to test that it doesn't throw?

File: Tests/Source/TransportSession/ZMTransportSessionTests.m [hunk 9]
```diff
@@ -240,6 +240,18 @@ - (instancetype)initWithScheduler:(ZMTransportRequestScheduler *)scheduler userA
...
 
+- (instancetype _Nonnull)initWithScheduler:(ZMTransportRequestScheduler * _Nonnull)scheduler
+                           userAgentString:(NSString * _Nonnull)userAgentString
+                               environment:(id<BackendEnvironmentProvider> _Nonnull)environment
+                             proxyUsername:(NSString * _Nullable)proxyUsername
+                             proxyPassword:(NSString * _Nullable)proxyPassword queue:(NSOperationQueue * _Nonnull)queue {
+    self = [self initWithScheduler:scheduler userAgentString:userAgentString environment:environment queue:queue];
+    if (self) {
+        // do more here if needed
+    }
```
Comment (David-Henner) [inline Tests/Source/TransportSession/ZMTransportSessionTests.m:251]: is this intended to stay?

File: Tests/Source/URLSession/BackendEnvironmentTests.swift [hunk 10]
```diff
@@ -55,16 +59,28 @@ class BackendEnvironmentTests: XCTestCase {
...
-        
-        guard let environment = BackendEnvironment(userDefaults: defaults, configurationBundle: backendBundle) else { XCTFail("Could not read environment data from Backend.bundle"); return }
+
+        guard
+            let environment = BackendEnvironment(
+            userDefaults: defaultsProd,
+            configurationBundle: backendBundle
+        ) else {
+            XCTFail("Could not read environment data from Backend.bundle"); return
+        }
```
Comment (David-Henner) [inline Tests/Source/URLSession/BackendEnvironmentTests.swift:83]:
```suggestion
        guard let environment = BackendEnvironment(
            userDefaults: defaultsProd,
            configurationBundle: backendBundle
        ) else {
            XCTFail("Could not read environment data from Backend.bundle"); return
        }
```

File: Tests/Source/URLSession/BackendEnvironmentTests.swift [hunk 11]
```diff
@@ -74,67 +90,117 @@ class BackendEnvironmentTests: XCTestCase {
...
+            XCTFail("Could not find configuration for custom")
+            return
+        }
+
+        guard
+            let data = try? Data(contentsOf: URL(fileURLWithPath: path))
+        else {
+                XCTFail("Could not read \(path)")
+                return
+        }
```
Comment (David-Henner) [inline Tests/Source/URLSession/BackendEnvironmentTests.swift:109]:
```suggestion
        guard
            let data = try? Data(contentsOf: URL(fileURLWithPath: path))
        else {
            XCTFail("Could not read \(path)")
            return
        }
```

File: Tests/Source/URLSession/BackendEnvironmentTests.swift [hunk 12]
```diff
@@ -74,67 +90,117 @@ class BackendEnvironmentTests: XCTestCase {
...
+
+        guard
+            let environment = BackendEnvironment(
+                environmentType: EnvironmentType(userDefaults: defaultsCustom),
+                data: data)
+        else {
```
Comment (David-Henner) [inline Tests/Source/URLSession/BackendEnvironmentTests.swift:115]:
```suggestion
        guard let environment = BackendEnvironment(
            environmentType: EnvironmentType(userDefaults: defaultsCustom),
            data: data
        ) else {
```

File: Source/Public/BackendEnvironment+Defaults.swift [hunk 13]
```diff
@@ -52,17 +53,31 @@ extension BackendEnvironment {
...
+            let proxy: ProxySettings?
+            if let proxySettingsFromProvider = proxySettings {
+                proxy = ProxySettings(
+                    host: proxySettingsFromProvider.host,
+                    port: proxySettingsFromProvider.port,
+                    needsAuthentication: proxySettingsFromProvider.needsAuthentication
+                )
+            } else {
+                proxy = nil
+            }
```
[the remark this answers is not in the document]
Comment (netbe) [reply Source/Public/BackendEnvironment+Defaults.swift:78]: cannot be used if not initialized, or is it for the local variable name?

File: Source/TransportSession/UnauthenticatedTransportSession.swift (hunk 4 above)
Comment (netbe) [reply Source/TransportSession/UnauthenticatedTransportSession.swift:107]: I'll double check but it was because of ObjC

File: Source/TransportSession/ZMTransportSession.m (hunk 6 above)
Comment (netbe) [reply Source/TransportSession/ZMTransportSession.m:185]: removed the if != nil as it has no effect

File: Tests/Source/ProxyCredentialsTests.swift (hunk 8 above)
Comment (netbe) [reply Tests/Source/ProxyCredentialsTests.swift]: yes that it does not throw, added assertion

File: Tests/Source/TransportSession/ZMTransportSessionTests.m (hunk 9 above)
Comment (netbe) [reply Tests/Source/TransportSession/ZMTransportSessionTests.m:251]: removed

File: Source/Public/BackendEnvironment+Defaults.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (David-Henner) [reply Source/Public/BackendEnvironment+Defaults.swift]: I edited the suggestion to set `proxy` to nil explicitly, would it work like that?
also yeah the variable name is just a suggestion to make it shorter

**Review by David-Henner — APPROVED** [review]

**Review by johnxnguyen — APPROVED** [review]

**Comment by netbe** [comment]
release major

**Comment by netbe** [comment]
release major

**Comment by zenkins** [comment]
Hello, @netbe! Merging PR for command **release major** for PR #**258**. Will create a **major** release after merged.
Next version is 92.0.0.

**Comment by netbe** [comment]
release major

**Comment by zenkins** [comment]
Hello, @netbe! Merging PR for command **release major** for PR #**258**. Will create a **major** release after merged.
Next version is 92.0.0.
